libs/core/coroutines/include/hpx/coroutines/thread_id_type.hpp

```diff
@@ -56,60 +74,277 @@ namespace hpx { namespace threads {
+
+        explicit thread_id(thread_id_repr const& thrd) noexcept
+          : thrd_(thrd)
+        {
+        }
+        explicit thread_id(thread_id_repr&& thrd) noexcept
+          : thrd_(std::move(thrd))
+        {
+        }
+
+        thread_id& operator=(thread_id_repr const& rhs) noexcept
+        {
+            thrd_ = rhs;
+            return *this;
+        }
+        thread_id& operator=(thread_id_repr&& rhs) noexcept
+        {
+            thrd_ = std::move(rhs);
+            return *this;
+        }
+
+        using thread_repr = detail::thread_data_reference_counting;
+
+        explicit thread_id(thread_repr* thrd, bool addref = true) noexcept
+          : thrd_(thrd, addref)
+        {
+        }
+
+        thread_id& operator=(thread_repr* rhs) noexcept
+        {
+            thrd_.reset(rhs);
+            return *this;
+        }
+
+        thread_id(thread_id_noref const& noref)
+          : thrd_(static_cast<thread_repr*>(noref.get()))
+        {
+        }
+
+        thread_id(thread_id_noref&& noref) noexcept
+          : thrd_(static_cast<thread_repr*>(noref.get()))
+        {
+            noref.reset();
+        }
+
+        thread_id& operator=(thread_id_noref const& noref)
+        {
+            thrd_.reset(static_cast<thread_repr*>(noref.get()));
+            return *this;
+        }
+
+        thread_id& operator=(thread_id_noref&& noref) noexcept
+        {
+            thrd_.reset(static_cast<thread_repr*>(noref.get()));
+            noref.reset();
+            return *this;
+        }
+
+        explicit operator bool() const noexcept
+        {
+            return nullptr != thrd_;
+        }
+
+        thread_id_repr& get() & noexcept
+        {
+            return thrd_;
+        }
+        thread_id_repr&& get() && noexcept
+        {
+            return std::move(thrd_);
+        }
+
+        thread_id_repr const& get() const& noexcept
+        {
+            return thrd_;
+        }
+
+        void reset() noexcept
+        {
+            thrd_.reset();
+        }
+
+        friend bool operator==(std::nullptr_t, thread_id const& rhs) noexcept
+        {
+            return nullptr == rhs.thrd_;
+        }
+
+        friend bool operator!=(std::nullptr_t, thread_id const& rhs) noexcept
+        {
+            return nullptr != rhs.thrd_;
+        }
+
+        friend bool operator==(thread_id const& lhs, std::nullptr_t) noexcept
+        {
+            return nullptr == lhs.thrd_;
+        }
+
+        friend bool operator!=(thread_id const& lhs, std::nullptr_t) noexcept
+        {
+            return nullptr != lhs.thrd_;
+        }
+
+        friend bool operator==(
+            thread_id const& lhs, thread_id const& rhs) noexcept
+        {
+            return lhs.thrd_ == rhs.thrd_;
+        }
+
+        friend bool operator!=(
+            thread_id const& lhs, thread_id const& rhs) noexcept
+        {
+            return lhs.thrd_ != rhs.thrd_;
+        }
+
+        friend bool operator<(
```

Comment:
I've nothing against this, but is this useful?

Reply:
It's needed if somebody wants to hold this type in a `std::set` or `std::map`, etc.